controllers/backstage_service.go

```diff
@@ -23,46 +23,64 @@ import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+func (r *BackstageReconciler) reconcileBackstageService(ctx context.Context, backstage bs.Backstage, ns string) error {
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("backstage-%s", backstage.Name),
+			Namespace: ns,
+		},
+	}
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, service, r.serviceObjectMutFun(ctx, service, backstage, ns)); err != nil {
+		if errors.IsConflict(err) {
+			return fmt.Errorf("retry sync needed: %v", err)
+		}
+		return err
+	}
+	return nil
+}
+
 // selector for deploy.spec.template.spec.meta.label
 // targetPort: http for deploy.spec.template.spec.containers.ports.name=http
-func (r *BackstageReconciler) applyBackstageService(ctx context.Context, backstage bs.Backstage, ns string) error {
+func (r *BackstageReconciler) serviceObjectMutFun(ctx context.Context, targetService *corev1.Service, backstage bs.Backstage, ns string) controllerutil.MutateFn {
+	return func() error {
+		service := &corev1.Service{}
+		targetService.ObjectMeta.DeepCopyInto(&service.ObjectMeta)
 
-	//lg := log.FromContext(ctx)
+		err := r.readConfigMapOrDefault(ctx, backstage.Spec.RawRuntimeConfig.BackstageConfigName, "service.yaml", ns, service)
+		if err != nil {
+			return err
+		}
 
-	service := &corev1.Service{}
-	err := r.readConfigMapOrDefault(ctx, backstage.Spec.RawRuntimeConfig.BackstageConfigName, "service.yaml", ns, service)
-	if err != nil {
-		return err
-	}
+		// Override the service name
+		service.Name = fmt.Sprintf("backstage-%s", backstage.Name)
+		setBackstageAppLabel(&service.Spec.Selector, backstage)
 
-	service.Name = fmt.Sprintf("backstage-%s", backstage.Name)
-	setBackstageAppLabel(&service.Spec.Selector, backstage)
+		r.labels(&service.ObjectMeta, backstage)
 
-	err = r.Get(ctx, types.NamespacedName{Name: service.Name, Namespace: ns}, service)
-	if err != nil {
-		if errors.IsNotFound(err) {
-		} else {
-			return fmt.Errorf("failed to get backstage service, reason: %s", err)
+		if r.OwnsRuntime {
+			if err := controllerutil.SetControllerReference(&backstage, service, r.Scheme); err != nil {
+				return fmt.Errorf("failed to set owner reference: %s", err)
+			}
 		}
-	} else {
-		//lg.Info("CR update is ignored for the time")
-		return nil
-	}
-
-	r.labels(&service.ObjectMeta, backstage)
 
-	if r.OwnsRuntime {
-		if err := controllerutil.SetControllerReference(&backstage, service, r.Scheme); err != nil {
-			return fmt.Errorf("failed to set owner reference: %s", err)
+		if len(targetService.Spec.ClusterIP) > 0 && service.Spec.ClusterIP != "" && service.Spec.ClusterIP != "None" && service.Spec.ClusterIP != targetService.Spec.ClusterIP {
+			return fmt.Errorf("backstage service IP can not be updated: %s, current: %s, new: %s", targetService.Name, targetService.Spec.ClusterIP, service.Spec.ClusterIP)
 		}
-	}
+		service.Spec.ClusterIP = targetService.Spec.ClusterIP
+		for _, ip1 := range targetService.Spec.ClusterIPs {
```

Comment:
When a service is created, normally k8s may update the spec to include the clusterIP and clusterIPs assigned randomly. These values can not be updated later.

Reply:
Well, it is often use case but not only, user is able to specify it or to set None for headless service
https://kubernetes.io/docs/concepts/services-networking/service/

Reply:
Yes the user is able to create headless service (actually the operator deploys a headless service for local db). What I meant here after the service has been created, the cluserIP should not be updated.

Reply:
Not sure I understand it.
Why we have this opinion and check it ourselves?

Reply:
Because without this checking, reconciliation would fail, as the operator tries to override the IPs that were populated by k8s, leading to an update error.

Reply:
It is not valid, I believe 
According to https://kubernetes.io/docs/concepts/services-networking/service/
If you do not specify ClusterIP (which is a case a believe) it is is assigned by K8s, i.e it is controlled be Kubernetes, Operator's desired state is still "I let K8s manage it" so, I believe operator must not try to override it.

Reply:
I do not think current implementation of update is correct, as discussed.

In general, an Operator must not deal with spec fields
"Types with both spec and status stanzas can (and usually should) have distinct authorization scopes for them. This allows users to be granted full write access to spec and read-only access to status, while relevant controllers are granted read-only access to spec but full write access to status."

And in this particular case, as stated https://kubernetes.io/docs/concepts/services-networking/cluster-ip-allocation/ if ClusterIP is dynamically assigned by control plane (the case we are testing) we should not manage it during CR update (specified desired state is ClusterIP=="", so it managed dynamically by K8s Control Plane).

So, I believe we have to revert this PR and rework this functionality.